File: python/cluster.py
import numpy as np
import sys
import ROOT

class cluster: # Only good for HD geometry 1x2x6
    def __init__(self, tps):
        self.tps_ = tps
        self.reco_pos_ = None
        self.n_tps_ = tps.shape[0]
    def __len__(self):
        return self.n_tps_
    def __str__(self):
        return str(self.tps_)
    def __repr__(self):
        return str(self.tps_)
    def get_tps(self):
        return self.tps_
    def get_reco_pos(self):
        return self.reco_pos_
    def set_variables(self, reco_pos = np.array([0,0,0])):
        self.reco_pos_ = reco_pos


def read_root_file_to_clusters(filename):
        # Create a structured array with column names
    dt = np.dtype([('time_start', float), 
                    ('time_over_threshold', float),
                    ('time_peak', float),
                    ('channel', int),
                    ('adc_integral', int),
                    ('adc_peak', int),
                    ('detid', int),
                    ('type', int),
                    ('algorithm', int),
                    ('version', int),
                    ('flag', int)])

    file = ROOT.TFile.Open(filename, "READ")
    clusters = []
    for i in file.GetListOfKeys():
        # get the TTree
        tree = file.Get(i.GetName())
        print(f"Tree name: {tree.GetName()}")
        tree.Print()
        counter = 0
        for entry in tree:
            counter += 1
            if counter % 1000 == 0:
                print(f"cluster {counter}")
            nrows = entry.nrows
            # m = np.empty((nrows), dtype=dt)
            # for j in range(nrows):
            #     m[j]["time_start"] = entry.matrix[j][0]
            #     m[j]["time_over_threshold"] = entry.matrix[j][1]
            #     m[j]["time_peak"] = entry.matrix[j][2]
            #     m[j]["channel"] = entry.matrix[j][3]
            #     m[j]["adc_integral"] = entry.matrix[j][4]
            #     m[j]["adc_peak"] = entry.matrix[j][5]
            #     m[j]["detid"] = entry.matrix[j][6]
            #     m[j]["type"] = entry.matrix[j][7]
            #     m[j]["algorithm"] = entry.matrix[j][8]
            #     m[j]["version"] = entry.matrix[j][9]
            #     m[j]["flag"] = entry.matrix[j][10]

            m = np.empty((nrows, 11))
            for j in range(nrows):
                m[j, 0] = entry.matrix[j][0]
                m[j, 1] = entry.matrix[j][1]
                m[j, 2] = entry.matrix[j][2]
                m[j, 3] = entry.matrix[j][3]
                m[j, 4] = entry.matrix[j][4]
                m[j, 5] = entry.matrix[j][5]
                m[j, 6] = entry.matrix[j][6]
                m[j, 7] = entry.matrix[j][7]
                m[j, 8] = entry.matrix[j][8]
                m[j, 9] = entry.matrix[j][9]
                m[j, 10] = entry.matrix[j][10]
            this_cluster = cluster(m)
            reco_pos = np.array([entry.reco_pos_x, entry.reco_pos_y, entry.reco_pos_z])
            this_cluster.set_variables(reco_pos = reco_pos)
            clusters.append(this_cluster)
        break
    return clusters

File: src/cluster.cpp
#include <iostream>
#include <vector>

#include "cluster.h"
#include "position_calculator.h" 

std::map<std::string, int> variables_to_index = {
    {"time_start", 0},
    {"time_over_threshold", 1},
    {"time_peak", 2},
    {"channel", 3},
    {"adc_integral", 4},
    {"adc_peak", 5},
    {"detid", 6},
    {"type", 7},
    {"algorithm", 8},
    {"version", 9},
    {"flag", 10},
};


void cluster::update_cluster_info() {
    float total_charge = 0;
    std::vector<float> reco_pos = {0, 0, 0};

    for (int i = 0; i < tps_.size(); i++) {
        total_charge += tps_[i][variables_to_index["adc_integral"]];
        std::vector<float> pos = calculate_position(tps_[i]);
        reco_pos[0] += pos[0];
        reco_pos[1] += pos[1];
        reco_pos[2] += pos[2];
    }

    total_charge_ = total_charge;
    int ntps = tps_.size();
    reco_pos[0] /= ntps;
    reco_pos[1] /= ntps;
    reco_pos[2] /= ntps;
    reco_pos_ = reco_pos;
}




void write_clusters_to_root(std::vector<cluster>& clusters, std::string root_filename) {
    // create folder if it does not exist
    std::string folder = root_filename.substr(0, root_filename.find_last_of("/"));
    std::string command = "mkdir -p " + folder;
    system(command.c_str());
    // create the root file
    TFile *f = new TFile(root_filename.c_str(), "recreate");
    TTree *tree = new TTree("tree", "tree");
    // prepare objects to save the data
    // std::vector<std::vector<int>> matrix;
    std::vector<std::vector<double>> matrix;
    int nrows;
    int reco_pos_x; 
    int reco_pos_y;
    int reco_pos_z;


    // create the branches
    tree->Branch("matrix", &matrix);
    tree->Branch("nrows", &nrows);
    tree->Branch("reco_pos_x", &reco_pos_x);
    tree->Branch("reco_pos_y", &reco_pos_y);
    tree->Branch("reco_pos_z", &reco_pos_z);

    // fill the tree
    for (auto& g : clusters) {
        matrix = g.get_tps();
        nrows = g.get_size();
        reco_pos_x = g.get_reco_pos()[0];
        reco_pos_y = g.get_reco_pos()[1];
        reco_pos_z = g.get_reco_pos()[2];
        tree->Fill();
    }
    // write the tree
    tree->Write();
    f->Close();

    return;   
}
